Adding persistent volume claims to our DB templates

mnagy opened the pull request:
Only MySQL for now, I'd appreciate early feedback before finishing the rest.
I've tested the MySQL image with NFS world-writable volumes and have verified
that the database works after scaling RC to 0 replicas and then back to 1.

cc @bparees @markturansky @rhcarvalho

mfojtik commented on examples/db-templates/mysql-template.json:
```diff
@@ -40,6 +40,23 @@
+        ],
+        "resources": {
+          "requests": {
+            "storage": "1Gi"
```
make this parameter

mnagy replied:
You beat me to it :) Question: Do we have some validation mechanisms for parameters? E.g. enforcing the correct format for storage capacity specification.

mfojtik replied:
if you specify wrong value, the List created from this template will fail to create, I don't think we need more validation

mnagy commented:
I wonder if the storage amount should be parametrizable? And I'd appreciate suggestions on the PVC naming, not sure if it's okay the way it is. I know it's bikeshedding a bit, but I think it's important to set good conventions.

mnagy commented:
Parametrized the capacity, made it 512Mi. PTAL

FYI, if anyone's interested in testing this (and for the purpose of future reference for me), this is my setup:
```
[fedora@ip-10-166-135-214 origin]$ cat /etc/exports
/nfs/01		*(rw,no_root_squash)
/nfs/02		*(rw,no_root_squash)
[fedora@ip-10-166-135-214 origin]$ ls -l /nfs
total 20
drwxrwxrwx. 2 root root 4096 Jun  1 14:56 01
drwxrwxrwx. 2 root root 4096 Jun  1 14:56 02
[fedora@ip-10-166-135-214 origin]$ cat persistent.yaml 
apiVersion: v1beta3
kind: List
items:
- apiVersion: v1beta3
  kind: PersistentVolume
  metadata:
    name: pv01
  spec:
    capacity:
      storage: 2Gi
    accessModes:
      - ReadWriteMany
    nfs:
      path: /nfs/01
      server: 172.17.42.1
- apiVersion: v1beta3
  kind: PersistentVolume
  metadata:
    name: pv02
  spec:
    capacity:
      storage: 2Gi
    accessModes:
      - ReadWriteMany
    nfs:
      path: /nfs/02
      server: 172.17.42.1
```

Also, needs https://github.com/GoogleCloudPlatform/kubernetes/pull/9055

sosiouxme commented:
Should this maybe be a separate example? Not everyone will have the backing PersistentVolume handy and might just want a test MySQL instance...

bparees commented:
yes we definitely want 2 distinct templates, one being what we already had (no volume, easy to run as a quick test drive) and the other being a prod-ready version that uses an NFS mount.

mnagy commented:
What about template names? mysql and mysql-persistent, or mysql and mysql-{test,example,something_else}? I think "mysql" being the one that's persistent is a better choice.

bparees commented:
i like mysql and mysql-persistent.

rhcarvalho commented:
How about mysql-nfs? NFS is one of the ways to make it persistent. There
could be other templates like mysql-aws, mysql-gce, etc.
On Jun 1, 2015 8:02 PM, "Ben Parees" <notifications@github.com> wrote:

> i like mysql and mysql-persistent.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/2644#issuecomment-107656308>.
>

mnagy commented:
@rhcarvalho you don't specify the storage method in the PV claim, you only request it to be persistent and to have certain capacity.

mnagy commented:
After off-github discussion with @bparees, decided to use `-persistent` and `-ephemeral`. Also added PostgreSQL, it works now as well.

bparees commented on examples/db-templates/mysql-persistent-template.json:
```diff
@@ -0,0 +1,192 @@
+      "name": "PVC_NAME",
+      "description": "persistent volume claim name",
+      "generate": "expression",
+      "from": "pvc-[a-z0-9]{8}"
```
how does this random generation work with claims?  is any random value valid as a claim id?

markturansky replied:
This needs to be the name of a claim that exists in the same namespace.  If pod can't find the claim by name, it will fail.

bparees replied:
so it doesn't seem like we should be generating random values then....

mnagy replied:
It needs to be random generated so it won't collide with other claims. While the name is generated, it is used consistently: We use the name to create a Persistent Volume Claim and also use it as a reference in container spec, to claim the volume.
FWIW, I tested the setup and the pods can find the volumes and will work correctly when RC is scaled.

bparees replied:
ok, @markturansky 's answer confused me but now i think i understand...  we are defining the claim with a random id, and then referencing that same claim.  Defining the claim with a random id is evidently valid.

markturansky replied:
Sorry to cause confusion.   If the claim name is generated upfront and that name is re-used in the pod/volume, all is well.

smarterclayton replied:
Why would my database claim name not match my database name?  "sampledb-data"?  The whole point of names is to uniquely identify things.  The PVC name should match the name of the things using it in most cases

----- Original Message -----
> > +    },
> > +    {
> > +      "name": "MYSQL_DATABASE",
> > +      "description": "database name",
> > +      "value": "sampledb"
> > +    },
> > +    {
> > +      "name": "PVC_CAPACITY",
> > +      "description": "database volume capacity, e.g. 512Mi, 2Gi",
> > +      "value": "512Mi"
> > +    },
> > +    {
> > +      "name": "PVC_NAME",
> > +      "description": "persistent volume claim name",
> > +      "generate": "expression",
> > +      "from": "pvc-[a-z0-9]{8}"
> 
> Sorry to cause confusion.   If the claim name is generated upfront and that
> name is re-used in the pod/volume, all is well.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2644/files#r31526277

mfojtik replied:
@smarterclayton we want to make it random to avoid collisions... how about `sampledb-data-#####`

liggitt replied:
+1 for inspectable prefix + random suffix

mnagy replied:
Prefix + random suffix seems best. But yeah, it can't be sampledb-data, that could conflict easily.

smarterclayton replied:
Collisions with what?

> On Jun 2, 2015, at 10:36 AM, Michal Fojtik <notifications@github.com> wrote:
> 
> In examples/db-templates/mysql-persistent-template.json:
> 
> > +    },
> > +    {
> > +      "name": "MYSQL_DATABASE",
> > +      "description": "database name",
> > +      "value": "sampledb"
> > +    },
> > +    {
> > +      "name": "PVC_CAPACITY",
> > +      "description": "database volume capacity, e.g. 512Mi, 2Gi",
> > +      "value": "512Mi"
> > +    },
> > +    {
> > +      "name": "PVC_NAME",
> > +      "description": "persistent volume claim name",
> > +      "generate": "expression",
> > +      "from": "pvc-[a-z0-9]{8}"
> @smarterclayton we want to make it random to avoid collisions... how about sampledb-data-#####
> 
> —
> Reply to this email directly or view it on GitHub.
>

smarterclayton replied:
Who is it conflicting with?

> On Jun 2, 2015, at 10:50 AM, Martin Nagy <notifications@github.com> wrote:
> 
> In examples/db-templates/mysql-persistent-template.json:
> 
> > +    },
> > +    {
> > +      "name": "MYSQL_DATABASE",
> > +      "description": "database name",
> > +      "value": "sampledb"
> > +    },
> > +    {
> > +      "name": "PVC_CAPACITY",
> > +      "description": "database volume capacity, e.g. 512Mi, 2Gi",
> > +      "value": "512Mi"
> > +    },
> > +    {
> > +      "name": "PVC_NAME",
> > +      "description": "persistent volume claim name",
> > +      "generate": "expression",
> > +      "from": "pvc-[a-z0-9]{8}"
> Prefix + random suffix seems best. But yeah, it can't be sampledb-data, that could conflict easily.
> 
> —
> Reply to this email directly or view it on GitHub.
>

mnagy replied:
With other PVCs.

smarterclayton replied:
What other pvcs?

> On Jun 2, 2015, at 10:52 AM, Martin Nagy <notifications@github.com> wrote:
> 
> In examples/db-templates/mysql-persistent-template.json:
> 
> > +    },
> > +    {
> > +      "name": "MYSQL_DATABASE",
> > +      "description": "database name",
> > +      "value": "sampledb"
> > +    },
> > +    {
> > +      "name": "PVC_CAPACITY",
> > +      "description": "database volume capacity, e.g. 512Mi, 2Gi",
> > +      "value": "512Mi"
> > +    },
> > +    {
> > +      "name": "PVC_NAME",
> > +      "description": "persistent volume claim name",
> > +      "generate": "expression",
> > +      "from": "pvc-[a-z0-9]{8}"
> With other PVCs.
> 
> —
> Reply to this email directly or view it on GitHub.
>

smarterclayton replied:
A PVC is either 1:1 with the RC (because it's a singleton) or it's 1:N (because we expect scale up).  If it's scale up, then we should be using generateName on the PVC with a common prefix with the RC/DC

mnagy replied:
@smarterclayton The images don't support replication, so the former case is relevant. Is there any way to forbid RC to go higher than 1 replica? Or do we just tell it to the user through templates description?

mfojtik replied:
@mnagy there is no way to forbid users doing this (we should state it clearly in template description/annotation).

bparees replied:
we have an annotation on the images themselves that suggest the image should not be scaled but i guess our ui components don't expose that.

liggitt replied:
can you open an issue for that with the annotation called out?

liggitt replied:
(although, I'm not sure the UI goes and fetches the images at this point... it might just use the info in the template)

smarterclayton replied:
The unique persistent volume claim will prevent multiple replicas from starting, which is exactly the behavior we want.

----- Original Message -----
> > +    },
> > +    {
> > +      "name": "MYSQL_DATABASE",
> > +      "description": "database name",
> > +      "value": "sampledb"
> > +    },
> > +    {
> > +      "name": "PVC_CAPACITY",
> > +      "description": "database volume capacity, e.g. 512Mi, 2Gi",
> > +      "value": "512Mi"
> > +    },
> > +    {
> > +      "name": "PVC_NAME",
> > +      "description": "persistent volume claim name",
> > +      "generate": "expression",
> > +      "from": "pvc-[a-z0-9]{8}"
> 
> @smarterclayton The images don't support replication, so the former case is
> relevant. Is there any way to forbid RC to go higher than 1 replica? Or do
> we just tell it to the user through templates description?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2644/files#r31628249

bparees replied:
actually i'm not seeing the label on our images.  @mfojtik did we not end up doing that, or did you remove them when you made the DB images scalable?

bparees commented:
@mnagy I spoke with @smarterclayton and @danmcp about this a bit and the conclusion we reached is that we should add EmptyDir volumes to the ephemeral templates.

longer term we also want to add metadata to our images identifying volumes that need persistent storage.

For OSE we will then package just the persistent templates.  The ephemeral ones can be used by anyone wishing to quickly demo openshift or setting up an environment where DB persistence is not important.

mnagy commented:
@bparees can't use EmptyDir. The mounted directory is owned by root.root with permissions set to 750.

bparees commented:
@mnagy that would make emptydir useless...  @smarterclayton  ??

smarterclayton commented:
I've bumped that to a p0 blocking issue on 3.0 - tracked under #2233.  I need an owner to shepherd it, that's @pmorie at the moment.

----- Original Message -----
> @mnagy that would make emptydir useless...  @smarterclayton  ??
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2644#issuecomment-108524208

mnagy commented:
Updated. Added mandatory mongo admin password and changed parameter descriptions, template descriptions as well.

@smarterclayton with these templates, I see that if I try to scale to two replicas a new pod is created and tries to access the same PVC. In case of mongo at least, it just fails, but RC tries to redeploy. Am I doing something wrong or is that a bug? The PVC name is unique.

derekwaynecarr commented:
@markturansky - see above question on PVC from prior comment.

markturansky commented:
@mnagy For each pod to have its own volume, one is forced to use a ReplicationController for each with replica count = 1.

A PVC is bound to exactly 1 PV.   To reuse that 1 PV across many pod replicas, you have have to use a volume that supports many mounts (like GCEPD can mount 16 in ReadOnly mode or NFS can mount many in Read/Write mode).

If you want to use a distinct volume per mongo node, you would need X claims for X nodes.

mnagy commented:
@markturansky The problem I have is that if I scale what I have now, the PVC is reused. Am I doing something wrong? Can you please take a look at mongodb-persistent-template.json in this PR?

markturansky commented:
Scaling mongo would require making a new PVC for each PV you want to use.  Want 10 mongo nodes?  You would need 10 claims.   A Template uses 1 claim, so you would need 10 templates.

I understand that you'd like mongo to get a new PV for each node and use the same claim in the Template.  That is, unfortunately, not how it works.  That problem is more complex and was deliberately left out of the first version of PV.

smarterclayton commented:
The PVC being reused now isn't a problem.  The problem would be if the nodes aren't properly blocking your startup because the PVC is already bound somewhere else.

----- Original Message -----
> @markturansky The problem I have is that if I scale what I have now, the PVC
> is reused. Am I doing something wrong? Can you please take a look at
> mongodb-persistent-template.json in this PR?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2644#issuecomment-108590546

markturansky commented:
PVC being bound isn't a problem, but mounted can be.  

If PV is already mounted somewhere, it may (or may not, depending on volume type) support being mounted elsewhere, too.  Validation of this kind does not exist.

smarterclayton commented:
Most won't.  The node needs to be able to understand that it can't reuse an existing mount like this - if you have PV used by pod 1 and pod 2, and pod 2 is scheduled onto the same host, pod 2 had better not be able to mount the volume.

----- Original Message -----
> PVC being bound isn't a problem, but mounted can be.
> 
> If PV is already mounted somewhere, it may (or may not, depending on volume
> type) support being mounted elsewhere, too.  Validation of this kind does
> not exist.
> 
> 
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2644#issuecomment-108593764

mnagy commented:
I don't want to scale them, the images are not ready for that yet.

The problem I have is, that if you accidentally scale mongo, the second pod will reuse the PVC, so you'll have two pods using the same PVC and hence having mounted the same volume. The pod will then fail and keep pending because mongo inside the container will not start because of a pid lock that's on the volume.

@smarterclayton I'm concerned over this only because of what you said earlier https://github.com/openshift/origin/pull/2644#discussion_r31643756 about uniqueness of PVC blocking the scaling, which is not the case. I was wondering if I'm doing something wrong in the template.

smarterclayton commented:
It won't block scaling, but it *should* block startup.

----- Original Message -----
> I don't want to scale them, the images are not ready for that yet.
> 
> The problem I have is, that if you accidentally scale mongo, the second pod
> will reuse the PVC, so you'll have two pods using the same PVC and hence
> having mounted the same volume. The pod will then fail and keep pending
> because mongo inside the container will not start because of a pid lock
> that's on the volume.
> 
> @smarterclayton I'm concerned over this only because of what you said earlier
> https://github.com/openshift/origin/pull/2644#discussion_r31643756 about
> uniqueness of PVC blocking the scaling, which is not the case. I was
> wondering if I'm doing something wrong in the template.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2644#issuecomment-108596230

markturansky commented:
So mongo is failing correctly?  It fails in the expected way when giving the same volume to mount, which we know is a known limitation in PV.

This would be the case for other volumes like EBS, where we expect the call to attach to fail in the AWS API because the volume is already mounted elsewhere.

Not every volume should block on startup if already mounted.  Some can.  Type matters.

mnagy commented:
@smarterclayton it doesn't. It retries over and over. Not a problem for mongo and shouldn't be problem for our other DB images, but might be for other images.

If what @markturansky says about the same volume mounting being a known limitation then we don't have an issue here, as far as I understand it.

@bparees PTAL

smarterclayton commented:
----- Original Message -----
> @smarterclayton it doesn't. It retries over and over. Not a problem for mongo
> and shouldn't be problem for our other DB images, but might be for other
> images.
> 

It doesn't what?  It should retry forever on the node, but only one extra pod should be created.  

> If what @markturansky says about the same volume mounting being a known
> limitation then we don't have an issue here, as far as I understand it.
> 
> @bparees PTAL
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2644#issuecomment-108601068

mnagy commented:
It doesn't block startup. The container in the new pod doesn't start just because the container entrypoint fails and exits. I might have misunderstand you earlier and this might be what we expect.

mfojtik commented:
Right, the container itself will keep retrying and you should notice that
in UI as the pod will change the color... Clayton, can we also add some
notification to osc? like exclamation mark....
On Jun 3, 2015 10:14 PM, "Clayton Coleman" <notifications@github.com> wrote:

>
>
> ----- Original Message -----
> > @smarterclayton it doesn't. It retries over and over. Not a problem for
> mongo
> > and shouldn't be problem for our other DB images, but might be for other
> > images.
> >
>
> It doesn't what? It should retry forever on the node, but only one extra
> pod should be created.
>
> > If what @markturansky says about the same volume mounting being a known
> > limitation then we don't have an issue here, as far as I understand it.
> >
> > @bparees PTAL
> >
> > ---
> > Reply to this email directly or view it on GitHub:
> > https://github.com/openshift/origin/pull/2644#issuecomment-108601068
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/2644#issuecomment-108601379>.
>

smarterclayton commented:
There was a plan for that to be added.

Here's what I expect to happen.

1. Dev creates PVC pointing to PV1
2. Dev creates RC scaled to 0 that references the PVC
3. Dev scales RC to 1.
4. Pod starts, PV is bound to node1, volume PV1 is mounted, code is happy
5. Dev scales RC to 2.
6. Pod is scheduled to node1 (randomly)
7. Pod should try to mount PV1 and fail. Container entrypoint should never be invoked.

----- Original Message -----
> Right, the container itself will keep retrying and you should notice that
> in UI as the pod will change the color... Clayton, can we also add some
> notification to osc? like exclamation mark....
> On Jun 3, 2015 10:14 PM, "Clayton Coleman" <notifications@github.com> wrote:
> 
> >
> >
> > ----- Original Message -----
> > > @smarterclayton it doesn't. It retries over and over. Not a problem for
> > mongo
> > > and shouldn't be problem for our other DB images, but might be for other
> > > images.
> > >
> >
> > It doesn't what? It should retry forever on the node, but only one extra
> > pod should be created.
> >
> > > If what @markturansky says about the same volume mounting being a known
> > > limitation then we don't have an issue here, as far as I understand it.
> > >
> > > @bparees PTAL
> > >
> > > ---
> > > Reply to this email directly or view it on GitHub:
> > > https://github.com/openshift/origin/pull/2644#issuecomment-108601068
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/openshift/origin/pull/2644#issuecomment-108601379>.
> >
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2644#issuecomment-108603295

markturansky commented:
@smarterclayton that is correct for some volumes but not all.   NFS can mount them all read/write and it would be correct for the RC to scale to 2.   This may or may not be the users intent when scaling to RC to 2, but it could be depending on the volume/application.   WordPress, for example, would work within this.

smarterclayton commented:
> On Jun 3, 2015, at 4:37 PM, Mark Turansky <notifications@github.com> wrote:
> 
> @smarterclayton that is correct for some volumes but not all. NFS can mount them all read/write and it would be correct for the RC to scale to 2. This may or may not be the users intent when scaling to RC to 2, but it could be depending on the volume/application. WordPress, for example, would work within this.
> 
Sure, I just want to be totally sure our non-shared pvs don't reuse the existing mount incorrectly.  That would be a violation of the guarantee that volumes aren't reused.
> —
> Reply to this email directly or view it on GitHub.
>

markturansky commented:
@smarterclayton that makes sense and that type of smart validation does not yet exist.  We'll surely want that post kube 1.0 and it can probably happen with the same type of scheduling validation required (e.g can't put than 16 GCEPDs on a node)

smarterclayton commented:
So if the volume is already mounted, and has one pod using it, we're not defending against that?  Please open a defect against that - that's part of our responsibility to control and manage.  In practice we're unlikely to be scheduled on the same node, but it needs to be defended against.

----- Original Message -----
> @smarterclayton that makes sense and that type of smart validation does not
> yet exist.  We'll surely want that post kube 1.0 and it can probably happen
> with the same type of scheduling validation required (e.g can't put than 16
> GCEPDs on a node)
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2644#issuecomment-108608996

mnagy commented:
Just noticed I'm using ReadWriteMany for the PVC. Will ReadWriteOnce help?

markturansky commented:
@mnagy no.  that's another good area for better runtime validation of a claim (making sure the volume supports that type of access).

smarterclayton commented:
Link the tracking issue you create here mark and cc me on it.

----- Original Message -----
> @mnagy no.  that's another good area for better runtime validation of a claim
> (making sure the volume supports that type of access).
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2644#issuecomment-108615909

bparees commented on examples/examples_test.go:
```diff
@@ -82,9 +82,13 @@ func TestExampleObjectSchemas(t *testing.T) {
+			"mysql-persistent-template":      &templateapi.Template{},
+			"postgresql-persistent-template": &templateapi.Template{},
+			"mongodb-persistent-template":    &templateapi.Template{},
+			"mongodb-template":               &templateapi.Template{},
```
delete?

bparees commented:
@mnagy we definitely can't merge this until the emptydir issue is resolved, if nothing else.

markturansky commented:
@smarterclayton here is the new one:  https://github.com/GoogleCloudPlatform/kubernetes/issues/9218

Another issue for scheduling constraints is: https://github.com/GoogleCloudPlatform/kubernetes/issues/7835

These two are related and I can see 7835 growing the encompass the new one.

bparees commented:
@mnagy at this point i'm inclined to remove the emptydir so we can get this merged :/

smarterclayton commented:
Yes for now.  Going to have to be a follow up.

> On Jun 6, 2015, at 1:37 AM, Ben Parees <notifications@github.com> wrote:
> 
> @mnagy at this point i'm inclined to remove the emptydir so we can get this merged :/
> 
> —
> Reply to this email directly or view it on GitHub.
>

mnagy commented:
Removed emptyDir, PTAL.

mfojtik commented on examples/db-templates/mongodb-ephemeral-template.json:
```diff
@@ -128,23 +132,29 @@
     {
       "name": "MONGODB_DATABASE",
-      "description": "database name",
+      "description": "Database name.",
```
Remove the trailing dot.

mfojtik commented on examples/db-templates/mongodb-persistent-template.json:
```diff
@@ -0,0 +1,196 @@
+  "apiVersion": "v1beta3",
+  "metadata": {
+    "name": "mongodb-persistent",
+    "creationTimestamp": null,
```
nuke this

mfojtik commented on examples/db-templates/mongodb-persistent-template.json:
```diff
@@ -0,0 +1,196 @@
+  "parameters": [
+    {
+      "name": "MONGODB_USER",
+      "description": "Username for MongoDB user that will be used for accessing the database.",
```
same here, remove all trailing dots

mnagy replied:
Why do we not want full sentences in the UI?

liggitt replied:
It's already not a full sentence :)

mfojtik commented on examples/db-templates/postgresql-persistent-template.json:
```diff
@@ -0,0 +1,186 @@
+            ],
+            "restartPolicy": "Always",
+            "dnsPolicy": "ClusterFirst",
+            "serviceAccount": ""
```
nuke this (and all other occurrences)

mnagy commented:
Addressed all comments, PTAL

mfojtik commented:
LGTM (you have to get @smarterclayton approval to merge)

smarterclayton commented:
Approved, [merge]
